### PR TITLE
Ignore unreachable code in wasm binaries

### PR DESCRIPTION
**Pull request by kripken**
This proposes a change in direction for how we handle unreachable code when reading wasm binaries, that we simply ignore it. The corner cases that the fuzzer found in wasm unreachable code have turned out to be tricky, and getting them to work well in binaryen's IR just doesn't seem worth the effort.

We do still have a way to write and read code with unreachable parts, using `.wast` files, so this doesn't hurt our testing infrastructure. The only change here is when reading wasm binaries.

### CODE REVIEW

**Comment by dschuff** [comment]
Does this actually help anything? Can't optimizers, wast files, and clients of the Binaryen C interface create code that is just a problematic? Also, this means that wast files are no longer isomorphic or round-trippable with binary files, which seems bad. I know we have a split between wasm and Binaryen IR, but previously we still had the property that if you encode Binaryen IR to binary or wast and read it back, you'd get the same thing, which is no longer the case.

**Comment by kripken** [comment]
I don't think it's possible to create problems in wast text files - since we don't parse the stacky stuff. We just parse structured s-expr code, which doesn't run into problems. The only problem is in wasm binaries, which this PR works around.

About round-tripping, actually we already don't have binary round-tripping being identical, also because of unreachable stuff - we emit extra unreachables in some cases, for places we allow unreachable things but wasm doesn't, like blocks and ifs. But it's true that this PR takes us much further in that direction. In practice, I don't think it's too much of a problem - unreachable code won't be run so as an optimizer, binaryen doesn't care about it; and otherwise, when we need the ability to save IR in a lossless manner, we can use wasts (and some day we may add a binaryen IR binary format).

File: test/passes/flatten-control-flow.bin.txt [hunk 1]
```diff
@@ -119,51 +119,7 @@
   (local $var$8 f64)
   (block $label$0
    (nop)
-   (block
```
Comment (dschuff) [inline test/passes/flatten-control-flow.bin.txt]: Should this test be updated so we don't remove a big chunk of it?

File: README.md [hunk 2]
```diff
@@ -37,6 +37,7 @@ The differences between Binaryen IR and WebAssembly are:
  * Binaryen IR [is an AST](https://github.com/WebAssembly/binaryen/issues/663), for convenience of optimization. This differs from the WebAssembly binary format which is a stack machine.
  * WebAssembly limits block/if/loop types to none and the concrete value types (i32, i64, f32, f64). Binaryen IR has an unreachable type, and it allows block/if/loop to take it, allowing [local transforms that don't need to know the global context](https://github.com/WebAssembly/binaryen/issues/903).
  * Binaryen IR's text format requires the names of blocks and loops to be unique. This differs from the WebAssembly s-expression format which allows duplicate names (and depends on scoping to disambiguate).
+ * Binaryen ignores unreachable code when reading WebAssembly binaries (the code won't execute anyhow, and WebAssembly unreachable code has tricky corner cases that do not match up well against Binaryen IR).
```
Comment (dschuff) [inline README.md]: This paragraph should probably have a lot more detail about what the behavior is and why.

File: test/passes/flatten-control-flow.bin.txt (hunk 1 above)
Comment (kripken) [inline test/passes/flatten-control-flow.bin.txt]: Looks like the goal was to see that unreachable code doesn't break us, there was nothing in need of flattening after the unreachable. So the test is still fulfilling that role. Code truly in need of flattening is tested in $9 after this one, and in the wast (text) test for this.

File: README.md (hunk 2 above)
Comment (kripken) [inline README.md]: improved, let me know what you think.

Comment (dschuff) [inline README.md]: That part looks good, I think now it just needs a bit explaining the behavior difference for wast files.

Comment (kripken) [inline README.md]: Added.

Comment (dschuff) [inline README.md]: Looks good, thanks.

**Pull request merged**